### ISSUE TITLE
navigator.share is not always a function

### ISSUE THREAD

**Issue by roebuk** [microsoft/TypeScript#42809]
# Bug Report

navigator.share is undefined is many browsers, yet TypeScript reports it as "the function is always defined"

### 🔎 Search Terms

navigator.share
share

### 🕗 Version & Regression Information

Version 4.1.5 & Nightly

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/JYMwBAFAdghgbsA5jALgewE4DoDOALGDAUwEowBvAWACgw6wBjNKHNAGyKzbUQgHIAKniJhYKYHBH5CIkEVQBXYmGA4VAWwAOHdUSgoiAEz4kA3DQC+YImxwiqtekxbtO3XoOGjUEqQWVyisqqomgoGtpEuvpGJubUFkA

### 💻 Code

```ts
if (navigator.share) {
    console.log('The native share feature is implemented');
} else {
    console.log('The native share feature is not implemented');
}
```

### 🙁 Actual behavior

TypeScript reports it as "the function is always defined"

### 🙂 Expected behavior

If should report navigator.share as maybe being present.

**Comment by MartinJohns** [comment]
I'm surprised this definition is provided in the first place, given that it's only specified in a **draft** specification and only two browsers support it (Edge and Safari, not even Chrome or Firefox).

**Comment by IllusionMH** [comment]
There is ton of definitions from Edge Web IDLs so I think this definition is one from Edge.

**Comment by IllusionMH** [comment]
Apparently no - it's not from Edge.

https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/837
#18642

**Comment by pridemusvaire** [comment]
I have a workaround for now., set a state variable that I update when the component loads

`const [canShare, setCanShare] = React.useState<any | undefined>(undefined);`
`useEffect(()=> {setCanShare(navigator.share)}, [])`

**Comment by jnastaskin** [comment]
> Ran into the same issue. Discovered that using `navigator['share']` instead also works around the Typescript error, at least with my installed version of TypeScript 4.5.3.

This worked for me thank you!!

**Comment by gettersgetmore** [comment]
or `'share' in navigator`

**Comment by RyanCavanaugh** [comment]
`Navigator.share` is intentionally modeled as present once an API is in the relevant specification and supported by two major browsers. As [the DOM library policy explains](https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1486#issuecomment-1416182571), browser-specific availability is not represented by making individual APIs optional. The API-specific optionality proposal in [TypeScript-DOM-lib-generator#1748](https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/1748) was closed under that policy.

With `--strictNullChecks`, this condition reports TS2774 in TypeScript 4.1.5, 6.0.3, and current native TypeScript because `share(data?: ShareData): Promise<void>` is required in the DOM declarations. Use a runtime availability check such as `'share' in navigator` where the application needs to support browsers without the API.
